### PR TITLE
TOOL-1105 Updated caching for NuGet restore step.

### PR DESCRIPTION
**Pull request by richard-bogdan-bitrise**

### CODE REVIEW

File: Gopkg.toml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+  branch = "master"
+  name = "github.com/bitrise-io/go-steputils"
+
+[[constraint]]
+  branch = "master"
+  name = "github.com/bitrise-io/go-utils"
+
+[[constraint]]
+  name = "github.com/bitrise-io/go-xamarin"
+  version = "1.2.0"
```
Comment (godrei) [inline Gopkg.toml]: for now loc tooling maintained package on master branch

File: Gopkg.toml [hunk 2]
```diff
@@ -0,0 +1,46 @@
+# Gopkg.toml example
```
Comment (godrei) [inline Gopkg.toml]: pls remove the examples

File: main.go [hunk 3]
```diff
@@ -84,74 +78,51 @@ func DownloadFile(downloadURL, targetPath string) error {
 	return nil
 }
 
-func main() {
-	configs := createConfigsModelFromEnvs()
-
+// downloadNugetWithVersion downloads NuGet with the given version.
+func downloadNugetWithVersion(nuGetRestoreCmdArgs *[]string, nugetVersion string) error {
```
Comment (godrei) [inline main.go]: return the downloaded NuGet binary path, instead of passing and modifying `nuGetRestoreCmdArgs`

File: main.go [hunk 4]
```diff
@@ -162,47 +133,105 @@ func main() {
...
+	if exists, err := pathutil.IsPathExists(httpCachePth); err != nil {
+		log.Warnf("Failed to determine if path (%s) exists, error: %s", httpCachePth, err)
+	} else if exists {
+		nuGetCache.IncludePath(httpCachePth)
+	}
+}
 
-	nugetCache := cache.New()
+// getHTTPCachePath gets the path for the HTTP cache.
+func getHTTPCachePath() string {
```
Comment (godrei) [inline main.go]: do not use `get` prefix: https://golang.org/doc/effective_go.html#Getters

File: main.go [hunk 5]
```diff
@@ -162,47 +133,105 @@ func main() {
...
+	if os.Getenv(cacheEnvHTTP) != "" {
```
Comment (godrei) [inline main.go]: if pth := os.Getenv(cacheEnvHTTP); pth != "" {
  return pth 
}

File: main.go [hunk 6]
```diff
@@ -162,47 +133,105 @@ func main() {
...
+		collectGlobalCaches(&nuGetCache)
 	}
+	return nuGetCache.Commit()
+}
 
-	// Collecting caches
-	fmt.Println()
-	log.Infof("Collecting NuGet cache...")
+// collectHTTPCaches collects the HTTP cache.
+func collectHTTPCaches(nuGetCache *cache.Cache) {
```
Comment (godrei) [inline main.go]: return the http cache path instead of modifying a function parameter, pls avoid using functions with a side effect, it is hard to follow how the data flows.

File: main.go [hunk 7]
```diff
@@ -162,47 +133,105 @@ func main() {
...
+func getHTTPCachePath() string {
+	if os.Getenv(cacheEnvHTTP) != "" {
+		return os.Getenv(cacheEnvHTTP)
+	}
+	return filepath.Join(pathutil.UserHomeDir(), ".local", "share", "NuGet", "v3-cache")
+}
 
-	xamarinHomeCache := filepath.Join(pathutil.UserHomeDir(), ".local", "share", "Xamarin")
+// collectGlobalCaches collects the global package caches.
+func collectGlobalCaches(nuGetCache *cache.Cache) {
```
Comment (godrei) [inline main.go]: same as above there, other than that i do not see the point of wrapping a single line of function call in a separate function.

File: main.go [hunk 8]
```diff
@@ -162,47 +133,105 @@ func main() {
...
+	nuGetCache.IncludePath(getGlobalCachePath())
+}
 
-	if exists, err := pathutil.IsPathExists(xamarinHomeCache); err != nil {
-		log.Warnf("Failed to determine if path (%s) exists, error: %s", xamarinHomeCache, err)
-	} else if exists {
-		nugetCache.IncludePath(xamarinHomeCache)
+// getGlobalCachePath gets the path for the global cache.
+func getGlobalCachePath() string {
+	if os.Getenv(cacheEnvGlobal) != "" {
```
Comment (godrei) [inline main.go]: refer my above similar comment

File: main.go [hunk 9]
```diff
@@ -162,47 +133,105 @@ func main() {
...
+		return os.Getenv(cacheEnvGlobal)
 	}
+	return filepath.Join(pathutil.UserHomeDir(), ".nuget", "packages")
+}
 
+// collectLocalCaches collects the local caches.
+func collectLocalCaches(nuGetCache *cache.Cache) {
 	absProjectRoot, err := filepath.Abs(".")
```
Comment (godrei) [inline main.go]: why the current dir? the solution path is a step input, so we can not expect that the current dir, is where the project exists. Use the dir of the solution instead.

File: main.go [hunk 10]
```diff
@@ -162,47 +133,105 @@ func main() {
...
+// getGlobalCachePath gets the path for the global cache.
+func getGlobalCachePath() string {
+	if os.Getenv(cacheEnvGlobal) != "" {
+		return os.Getenv(cacheEnvGlobal)
 	}
+	return filepath.Join(pathutil.UserHomeDir(), ".nuget", "packages")
+}
 
+// collectLocalCaches collects the local caches.
+func collectLocalCaches(nuGetCache *cache.Cache) {
```
Comment (godrei) [inline main.go]: return the errors and let the caller decide how to handle them

File: main.go [hunk 11]
```diff
@@ -1,54 +1,48 @@
...
+	"time"
 )
 
 // ConfigsModel ...
 type ConfigsModel struct {
-	XamarinSolution string
-	NugetVersion    string
+	XamarinSolution string `env:"xamarin_solution,file"`
+	NuGetVersion    string `env:"nuget_version"`
+	AllowCaching    string `env:"allow_caching,opt[local,global,none]"`
```
Comment (godrei) [inline main.go]: use cache level: https://github.com/bitrise-steplib/bitrise-step-android-unit-test/blob/master/step.yml#L105

File: step.yml [hunk 12]
```diff
@@ -42,3 +40,19 @@ inputs:
...
+        Allows caching for the step.
+
+        'local' enables the caching the packages in the build directory.
+        'global' enables the caching of the global-packages folder, this is where NuGet installs any downloaded package.
+        'none' disables the caching for the step.
+
+        Please find more information about caching here:
+        https://docs.microsoft.com/en-us/nuget/consume-packages/managing-the-global-packages-and-cache-folders
+      value_options:
+      - "local"
```
Comment (godrei) [inline step.yml:59]: have option to cache both local and global packages

**Review by godrei — CHANGES_REQUESTED** [review]
pls check my notes

File: Gopkg.toml (hunk 2 above)
Comment (richard-bogdan-bitrise) [reply Gopkg.toml]: Removed.

File: Gopkg.toml (hunk 1 above)
Comment (richard-bogdan-bitrise) [reply Gopkg.toml]: Updated to refer to master branch.

File: main.go (hunk 11 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Renamed to cache level.

File: step.yml (hunk 12 above)
Comment (richard-bogdan-bitrise) [reply step.yml:59]: Added case. Note: updated Jira ticket, as this wasn't a requirement.

File: main.go (hunk 10 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Updated to return the errors.

File: main.go (hunk 4 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Removed prefix.

File: main.go (hunk 5 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Updated.

File: main.go (hunk 8 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Updated.

File: main.go (hunk 3 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Updated code.

File: main.go (hunk 9 above)
Comment (richard-bogdan-bitrise) [reply main.go]: This part is legacy, so I cannot answer the why. However it makes sense to change it, so updated code.

File: main.go (hunk 6 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Updated code. However, please note:
The code looked like this before the update
```
func collectCaches(allowCaching string) error {
 	nuGetCache := cache.New()
 	switch allowCaching {
 	case cacheInputNone:
 		return nil
 	case cacheInputlocal:
 		collectLocalCaches(&nuGetCache)
 	case cacheInputGlobal:
 		collectGlobalCaches(&nuGetCache)
 	}		 	}
 	return nuGetCache.Commit()
 }
```

Now it looks like this:
```
func collectCaches(cacheLevel string, basePth string) error {
	nuGetCache := cache.New()
	switch cacheLevel {
	case cacheInputNone:
		return nil
	case cacheInputlocal:
		localCaches, err := collectLocalCaches(basePth)
		if err != nil {
			log.Warnf("Error occurred while getting local cache. Error: %s", err)
		}
		for _, lcItem := range localCaches {
			nuGetCache.IncludePath(lcItem)
		}
	case cacheInputGlobal:
		nuGetCache.IncludePath(collectGlobalCaches())
	case cacheInputAll:
		localCaches, err := collectLocalCaches(basePth)
		if err != nil {
			log.Warnf("Error occurred while getting all cache. Error: %s", err)
		}
		for _, lcItem := range localCaches {
			nuGetCache.IncludePath(lcItem)
		}
		nuGetCache.IncludePath(collectGlobalCaches())
	}
	return nuGetCache.Commit()
}
```

It became much harder to read. I agree, that it is harder to follow the flow when you are modifying a function param, but not much and I think this won't be a problem, if you choose a very understandable method name and also add documentation. Also, according to good coding principles, methods should be very short, ideally just a few lines.

File: main.go (hunk 7 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Removed wrapper.

**Comment by richard-bogdan-bitrise** [comment]
Updated code, please check.

File: main.go (hunk 6 above)
Comment (godrei) [reply main.go]: Yeah it was easiert to read, but in general functions with side effects are problematic, not just the data flow, but also testing is much harder. 

I think you can both avoid modifying function parameters, and keep the code readble. What about this:

```
func collectCaches(cacheLevel string, basePth string) error {
	if cacheLevel == cacheInputNone {
		return nil
	}

	nuGetCache := cache.New()

	if cacheLevel == cacheInputlocal || cacheLevel == cacheInputAll {
		localCaches, err := collectLocalCaches(basePth)
		if err != nil {
			log.Warnf("Error occurred while getting local cache: %s", err)
		}
		for _, lcItem := range localCaches {
			nuGetCache.IncludePath(lcItem)
		}
	}

	if cacheLevel == cacheInputGlobal || cacheLevel == cacheInputAll {
		nuGetCache.IncludePath(collectGlobalCaches())
	}

	return nuGetCache.Commit()
}
```

File: main.go [hunk 13]
```diff
@@ -84,74 +81,50 @@ func DownloadFile(downloadURL, targetPath string) error {
 	return nil
 }
 
-func main() {
-	configs := createConfigsModelFromEnvs()
-
+// downloadNugetWithVersion downloads NuGet with the given version.
+func downloadNugetWithVersion(nuGetVersion string) ([]string, error) {
```
Comment (godrei) [inline main.go]: why do you return []string?
I expect a download function to download something and return its path.
Also, the function name and its parameters are stuttering, i would use `downloadNuGet(version string) ...`

Comment (richard-bogdan-bitrise) [reply main.go]: Renamed and refactored.

File: main.go (hunk 6 above)
Comment (richard-bogdan-bitrise) [reply main.go]: As these methods (collectLocalCaches, collectGlobalCaches) work with files, ideally they should be mocked for unit tests, and we should test the collectCaches method. For this reason I updated that code, but passing the cache as arg would not make testing harder in this case.

I decided to go with the switch statement, it is longer than this new form, but it is much more structured.

File: main.go [hunk 14]
```diff
@@ -162,47 +135,120 @@ func main() {
...
 	if err != nil {
-		log.Warnf("Cache collection skipped: failed to determine project root path.")
-	} else {
-		err := filepath.Walk(absProjectRoot, func(path string, f os.FileInfo, err error) error {
-			if f.IsDir() {
-				if f.Name() == "packages" {
-					nugetCache.IncludePath(path)
-					return io.EOF
-				}
+		return []string{}, fmt.Errorf("cache collection skipped: failed to determine project root path")
```
Comment (godrei) [inline main.go]: pls always include the `error` in the wrapped error.

File: main.go [hunk 15]
```diff
@@ -162,47 +135,120 @@ func main() {
...
+				return io.EOF
 			}
-			return nil
-		})
-		if err == io.EOF {
-			err = nil
 		}
+		return nil
+	}); err != nil && err != io.EOF {
+		return []string{}, fmt.Errorf("cache collection skipped: failed to determine cache paths. Error: %s", err)
```
Comment (godrei) [inline main.go]: pls update `cache collection skipped: failed to determine cache paths: %s` : https://golang.org/doc/effective_go.html#errors

File: main.go [hunk 16]
```diff
@@ -162,47 +135,120 @@ func main() {
...
+	if err := runRestoreCommand(nuGetRestoreCmdArgs); err != nil {
+		fail("NuGet restore failed, error: %s", err)
+	}
+
+	// Collecting caches
+	fmt.Println()
+	log.Infof("Collecting NuGet cache...")
+	caches := collectCaches(configs.CacheLevel, path.Dir(configs.XamarinSolution))
+	if err := caches.Commit(); err != nil {
+		log.Warnf("Cache collection skipped: failed to commit cache paths.")
```
Comment (godrei) [inline main.go]: include the error too

File: main.go (hunk 15 above)
Comment (godrei) [reply main.go]: pls apply it everywhere you log/compose an error

File: main.go [hunk 17]
```diff
@@ -162,47 +135,120 @@ func main() {
...
+			nuGetCache.IncludePath(lcItem)
+		}
+		nuGetCache.IncludePath(collectGlobalCaches())
+	}
+	return nuGetCache
+}
 
-	nugetCache := cache.New()
+// collectHTTPCaches collects the HTTP cache.
+func collectHTTPCaches() (string, error) {
```
Comment (godrei) [inline main.go]: unused function

File: step.yml (hunk 12 above)
Comment (godrei) [reply step.yml:59]: pls add to the step.yml too.

File: main.go [hunk 18]
```diff
@@ -84,74 +81,50 @@ func DownloadFile(downloadURL, targetPath string) error {
...
-	}
-
-	fmt.Println()
-	log.Infof("Restoring Nuget packages...")
-
-	nugetRestoreCmdArgs = append(nugetRestoreCmdArgs, "restore", configs.XamarinSolution)
-
-	if err := retry.Times(1).Try(func(attempt uint) error {
+// runRestoreCommand runs the restore command with the given args.
+func runRestoreCommand(nuGetRestoreCmdArgs []string) error {
```
Comment (godrei) [inline main.go]: NOTE: nuGetRestoreCmdArgs could be way shorter as the scope of the variable is short, something like `cmd []string`, with this change the command.Model variable needs to be renamed for `c` or similar.

**Review by godrei — CHANGES_REQUESTED** [review]
Pls check my new notes

File: main.go (hunk 14 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Added

File: main.go (hunk 16 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Added.

File: main.go (hunk 17 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Removed.

File: main.go (hunk 18 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Changed it to cmdArgs, so the other variable could remain cmd.

File: step.yml (hunk 12 above)
Comment (richard-bogdan-bitrise) [reply step.yml:59]: Added.

File: main.go [hunk 19]
```diff
@@ -56,12 +55,12 @@ func DownloadFile(downloadURL, targetPath string) error {
 		}
 	}()
 	if err != nil {
```
Comment (godrei) [inline main.go]: based on the docs: https://golang.org/pkg/os/#Create

```
If successful, methods on the returned File can be used for I/O
```
i/o opertaions like close can be used only if there was no error. I would suggest to call the defer only in case of successful open.

File: main.go [hunk 20]
```diff
@@ -75,7 +74,7 @@ func DownloadFile(downloadURL, targetPath string) error {
 
 	_, err = io.Copy(outFile, resp.Body)
 	if err != nil {
-		return fmt.Errorf("failed to download from (%s), error: %s", downloadURL, err)
+		return fmt.Errorf("failed to download from (%s): %s", downloadURL, err)
```
Comment (godrei) [inline main.go]: failed to copy

File: main.go [hunk 21]
```diff
@@ -73,136 +69,166 @@ func DownloadFile(downloadURL, targetPath string) error {
...
-			if err := DownloadFile(nugetURL, downloadPth); err != nil {
-				log.Errorf("Failed to download nuget, error: %s", err)
-				return err
-			}
-			return nil
-		}); err != nil {
-			log.Errorf("Failed to download nuget, error: %s", err)
-			os.Exit(1)
+		if err := DownloadFile(nuGetURL, downloadPth); err != nil {
+			log.Errorf("Failed to download NuGet: %s", err)
```
Comment (godrei) [inline main.go]: only log this if the retry is not the last one otherwise, both this function and the client will print this error.

File: main.go [hunk 22]
```diff
@@ -73,136 +69,166 @@ func DownloadFile(downloadURL, targetPath string) error {
...
-			os.Exit(1)
+			fail("Failed to create NuGet command: %s", err)
 		}
 
 		cmd.SetStdout(os.Stdout)
 		cmd.SetStderr(os.Stderr)
 
 		if err := cmd.Run(); err != nil {
-			log.Errorf("Restore failed, error: %s", err)
+			log.Errorf("Restore failed: %s", err)
```
Comment (godrei) [inline main.go]: similar as above, both the client and the last retry will print the same error.

File: main.go (hunk 21 above)
Comment (godrei) [reply main.go]: i would also suggest to print warning here, as this will not make the step failing

File: main.go [hunk 23]
```diff
@@ -73,136 +69,166 @@ func DownloadFile(downloadURL, targetPath string) error {
...
-	log.Infof("Collecting NuGet cache...")
-
-	nugetCache := cache.New()
+	})
+}
 
-	xamarinHomeCache := filepath.Join(pathutil.UserHomeDir(), ".local", "share", "Xamarin")
+// collectCaches collects the caches based on the config.
+// For more information about caches please read: https://docs.microsoft.com/en-us/nuget/consume-packages/managing-the-global-packages-and-cache-folders
+func collectCaches(cacheLevel string, basePth string) cache.Cache {
```
Comment (godrei) [inline main.go]: Now if there were an error we return an empty cache item and call a commit on it, even if it is empty.
Either make this function to return every path to cache and let the client create a cache object, and commit the path. Or solve the whole commit here.

I would vote for the first as this way it is easier to test the function.

File: main.go [hunk 24]
```diff
@@ -73,136 +69,166 @@ func DownloadFile(downloadURL, targetPath string) error {
...
 	if err != nil {
-		log.Warnf("Cache collection skipped: failed to determine project root path.")
-	} else {
-		err := filepath.Walk(absProjectRoot, func(path string, f os.FileInfo, err error) error {
-			if f.IsDir() {
-				if f.Name() == "packages" {
-					nugetCache.IncludePath(path)
-					return io.EOF
-				}
+		return []string{}, fmt.Errorf("cache collection skipped: failed to determine project root path: %s", err)
```
Comment (godrei) [inline main.go]: `cache collection skipped` -> `cache collection failed`

File: main.go [hunk 25]
```diff
@@ -73,136 +69,166 @@ func DownloadFile(downloadURL, targetPath string) error {
...
+				return io.EOF
 			}
-			return nil
-		})
-		if err == io.EOF {
-			err = nil
 		}
+		return nil
+	}); err != nil && err != io.EOF {
+		return []string{}, fmt.Errorf("cache collection skipped: failed to determine cache paths: %s", err)
```
Comment (godrei) [inline main.go]: `cache collection skipped` -> `cache collection failed`

File: main.go [hunk 26]
```diff
@@ -73,136 +69,166 @@ func DownloadFile(downloadURL, targetPath string) error {
...
+	if err := runRestoreCommand(nuGetRestoreCmdArgs); err != nil {
+		fail("NuGet restore failed: %s", err)
+	}
+
+	// Collecting caches
+	fmt.Println()
+	log.Infof("Collecting NuGet cache...")
+	caches := collectCaches(configs.CacheLevel, path.Dir(configs.XamarinSolution))
+	if err := caches.Commit(); err != nil {
+		log.Warnf("Cache collection skipped: failed to commit cache paths: %s", err)
```
Comment (godrei) [inline main.go]: `Cache collection skipped` -> `Cache collection failed`

**Review by godrei — CHANGES_REQUESTED** [review]
Pls check my notes

File: main.go (hunk 19 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Changed order of code.

File: main.go (hunk 20 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Changed.

File: main.go (hunk 21 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Updated.

File: main.go (hunk 22 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Updated code.

File: main.go (hunk 24 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Updated.

File: main.go (hunk 25 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Updated.

File: main.go (hunk 26 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Updated.

File: main.go (hunk 23 above)
Comment (richard-bogdan-bitrise) [reply main.go]: Updated.

**Review by godrei — APPROVED** [review]
